control-flow/index.md

```diff
@@ -1,6 +1,368 @@
+
+```
+function randomDelayedResponse(text, callback) {
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  setTimeout(() => {
+    callback(text);
+  }, timeOut * 10);
+}
+
+randomDelayedResponse(1, text => {
+  console.log(text);
+  randomDelayedResponse(2, text => {
+    console.log(text);
+    randomDelayedResponse(3, text => {
+      console.log(text);
+      randomDelayedResponse(4, text => {
+        console.log(text);
+      });
+    });
+  });
+}); // outputs "1 2 3 4"
+```
+
+Structuring these callbacks works. However this approach can create code becomes difficult to understand and maintain. Let's look at another way.
+
+## Promises
+
+To avoid this callback hell, we can use a different structure called _promises_. A promise is a function that returns a _resolved_ response which we can chain using `then`, or a _rejected_ response which we can `catch`. It follows this pattern:
+
+```
+runOutPromise() // returns a response
+  .then(response => {
+    return foo; // this could optionally be another promise
+  })
+  .then(foo => {
+    // they can be chained
+  })
+  .catch(error => {
+    // catch "reject" output here
+  });
+```
+
+We can rewrite our runner example using promises like so:
+
+```
+function randomDelayedResponse(text, callback) {
+  return new Promise((resolve, reject) => {
+    const timeOut = Math.floor(Math.random() * 10) + 1;
+    setTimeout(() => {
+      resolve(text); // Replacing the callback with "resolve"
+    }, timeOut * 10);
+  });
+}
+
+randomDelayedResponse(1)
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(2);
+  })
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(3);
+  })
+  .then(res => {
+    console.log(res);
+    return randomDelayedResponse(4);
+  })
+  .then(res => {
+    console.log(res);
+  })
+  .catch(error => {
+    // handle error
+  })
+// outputs "1 2 3 4"
+```
+
+Running the above code should output the correct order, albeit with random delays between each number.
+
+We aren't handling errors in this example, but if our function was to encounter an error we would `reject('some error')` and this would then be picked up by the `catch`.
+
+You can learn more about `promises` here (LINK)
+
+Promises remove a lot of nesting and give us easier to read code. However there is a third way to control the execution order of our code.
+
+## Async / Await
+
+The third approach is built on top of the existing _promises_ approach but makes it easier to reason with. With `async` and `await` we can write code that feels a lot more like the simple top-down code, by telling it to wait when we need it to. Let's rewrite our _who will win?_ example from above.
+
+```
+function randomDelayedResponse(text, callback) {
```

Comment:
Maybe cut this one out, to emphasis it's exactly the same as the promise version? Potentially rename it to `randomDelayedResponsePromise` or something?

Reply:
Will try cutting that and the other duplicated code. I'm in two minds though, as it feels like it would be good if we could have each code snippet be copy-paste friendly

Reply:
Fair point - also makes it longer, so closer to book length :D haha.